kv/index.go

```diff
@@ -0,0 +1,537 @@
+func (i *Index) Initialize(ctx context.Context, store Store) error {
+	if err := store.Update(ctx, func(tx Tx) error {
+		_, err := i.indexBucket(tx)
+		return err
+	}); err != nil {
+		return err
+	}
+
+	if store, ok := store.(IndexPopulatorStore); ok && store.AutoPopulateIndex() {
+		_, err := i.Populate(ctx, store)
+		return err
+	}
+
+	return nil
+}
+
+func (i *Index) indexBucket(tx Tx) (Bucket, error) {
+	return tx.Bucket(i.IndexBucket())
+}
+
+func (i *Index) sourceBucket(tx Tx) (Bucket, error) {
+	return tx.Bucket(i.SourceBucket())
+}
+
+func indexKey(foreignKey, primaryKey []byte) (newKey []byte, _ error) {
+	newKey = make([]byte, len(primaryKey)+len(foreignKey)+1)
+	copy(newKey, foreignKey)
+	newKey[len(foreignKey)] = '/'
+	copy(newKey[len(foreignKey)+1:], primaryKey)
+
+	return
+}
+
+func indexKeyParts(indexKey []byte) (fk, pk []byte, err error) {
+	// this function is called with items missing in index
+	parts := bytes.SplitN(indexKey, []byte("/"), 2)
+	if len(parts) < 2 {
+		return nil, nil, errors.New("malformed index key")
+	}
+
+	// parts are fk/pk
+	fk, pk = parts[0], parts[1]
+
+	return
+}
+
+// Insert creates a single index entry for the provided primary key on the foreign key.
+func (i *Index) Insert(tx Tx, foreignKey, primaryKey []byte) error {
+	newKey, err := indexKey(foreignKey, primaryKey)
+	if err != nil {
+		return err
+	}
+
+	bkt, err := i.indexBucket(tx)
+	if err != nil {
+		return err
+	}
+
+	return bkt.Put(newKey, primaryKey)
+}
+
+// Delete removes the foreignKey and primaryKey mapping from the underlying index.
+func (i *Index) Delete(tx Tx, foreignKey, primaryKey []byte) error {
+	newKey, err := indexKey(foreignKey, primaryKey)
+	if err != nil {
+		return err
+	}
+
+	bkt, err := i.indexBucket(tx)
+	if err != nil {
+		return err
+	}
+
+	return bkt.Delete(newKey)
+}
+
+// VisitFunc is called for each k, v byte slice pair from the underlying source bucket
+// which are found in the index bucket for a provided foreign key.
+type VisitFunc func(k, v []byte) error
+
+// Walk walks the source bucket using keys found in the index using the provided foreign key
+// given the index has been fully populated.
+func (i *Index) Walk(tx Tx, foreignKey []byte, visitFn VisitFunc) error {
+	// skip walking if configured to do so as the index
+	// is currently being used purely to write the index
+	if !i.canRead {
+		return nil
+	}
+
+	sourceBucket, err := i.sourceBucket(tx)
+	if err != nil {
+		return err
+	}
+
+	indexBucket, err := i.indexBucket(tx)
+	if err != nil {
+		return err
+	}
+
+	cursor, err := indexBucket.ForwardCursor(foreignKey,
+		WithCursorPrefix(foreignKey))
+	if err != nil {
+		return err
+	}
+
+	return indexWalk(cursor, sourceBucket, visitFn, func(fk, pk []byte) error {
+		// fail iteration when key not found for item in index
+		return fmt.Errorf("for key %v indexed by %v: %w", pk, fk, ErrKeyNotFound)
+	})
+}
+
+// Populate does a full population of the index using the provided IndexOnFunc.
```

Comment:
Could you clarify what `IndexOnFunc` is? I cannot see where `IndexOnFunc` is provided in this context.

Reply:
Ahh another bit of documentation fallout :+1:  . IndexSourceOnFunc used to be called IndexOnFunc. I added source for clarification.